main/src/ca/uwaterloo/flix/language/phase/Weeder.scala

```diff
@@ -707,6 +734,78 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
               }
           }
 
+        case ParsedAst.Expression.NewChannel(sp1, tpe, expOpt, sp2) =>
+          expOpt match {
+            case None =>
+              // Case 1: NewChannel takes no expression that states the channel size
+              val channelSize = WeededAst.Expression.Int32(lit = 0, mkSL(sp2, sp2))
+              WeededAst.Expression.NewChannel(Types.weed(tpe), channelSize, mkSL(sp1, sp2)).toSuccess
+            case Some(exp) =>
+              // Case 2: NewChannel takes an expression that states the channel size
+              exp match {
+                case ParsedAst.Expression.Lit(_, ParsedAst.Literal.Int32(_, sign, _, _), _) if sign =>
+                    IllegalChannelSize(mkSL(sp1, sp2)).toFailure
+                case _ =>
+                  visit(exp, unsafe) map {
+                    case e => WeededAst.Expression.NewChannel(Types.weed(tpe), e, mkSL(sp1, sp2))
+                  }
+              }
+          }
+
+        case ParsedAst.Expression.GetChannel(sp1, exp, sp2) =>
+          visit(exp,unsafe) map {
+            case e => WeededAst.Expression.GetChannel(e, mkSL(sp1, sp2))
+          }
+
+        case ParsedAst.Expression.PutChannel(exp1, exp2, sp2) =>
+          val sp1 = leftMostSourcePosition(exp1)
+          val loc = mkSL(sp1, sp2)
+          @@(visit(exp1, unsafe), visit(exp2, unsafe)) map {
+            case (e1, e2) => WeededAst.Expression.PutChannel(e1, e2, loc)
+          }
+
+        case ParsedAst.Expression.Spawn(sp1, fn, params, sp2) =>
+          val loc = mkSL(sp1, sp2)
+          val func = WeededAst.Expression.VarOrDef(Name.QName(sp1, Name.RootNS, fn, sp2), loc)
+
+          // TODO simplify this
```

Comment:
I am okay if you cannot simplify it further, but could you add some more comments describing each step?

Reply:
Especially add comment that describe the overall goal and shows the transformation at a high level, i.e.
spawn f(e1, e2, e3) -> ...